apollo-compiler/src/main/kotlin/com/apollostack/compiler/ir/Field.kt

```diff
@@ -56,4 +57,6 @@ data class Field(
   fun isNonScalar() = fields?.any() ?: false
 
   fun hasFragments() = fragmentSpreads?.any() ?: false
+
+  fun isOptional(): Boolean = isConditional || !methodResponseType().endsWith("!")
```

Comment:
The `!` is part of the schema, that defines a field as non-nullable. But in a query definition you can put `@skip` and `@include` directives on fields to make them conditional (usually based on a variable). So that means even if a field is non-null in the schema, it can still be skipped dynamically by a particular query and we need to account for that as well.